econml/_rlearner.py

```diff
@@ -0,0 +1,330 @@
+
+        .. math::
+            \\frac{1}{n} \\sum_{i=1}^n (Y_i - \\hat{E}[Y|X_i, W_i]\
+                                        - \\hat{\\theta}(X_i)\\cdot (T_i - \\hat{E}[T|X_i, W_i]))^2
+
+        If `sample_weight` is not None at fit time, then a weighted average is returned. If the outcome Y
+        is multidimensional, then the average of the MSEs for each dimension of Y is returned.
+    """
+
+    def __init__(self, model_y, model_t, model_final,
+                 discrete_treatment, n_splits, random_state):
+        class ModelNuisance:
+            """
+            Nuisance model fits the model_y and model_t at fit time and at predict time
+            calculates the residual Y and residual T based on the fitted models and returns
+            the residuals as two nuisance parameters.
+            """
+
+            def __init__(self, model_y, model_t):
+                self._model_y = clone(model_y, safe=False)
+                self._model_t = clone(model_t, safe=False)
+
+            def fit(self, Y, T, X=None, W=None, Z=None, sample_weight=None):
+                assert Z is None, "Cannot accept instrument!"
+                self._model_t.fit(X, W, T, sample_weight=sample_weight)
+                self._model_y.fit(X, W, Y, sample_weight=sample_weight)
+                return self
+
+            def predict(self, Y, T, X=None, W=None, Z=None, sample_weight=None):
+                Y_pred = self._model_y.predict(X, W)
+                T_pred = self._model_t.predict(X, W)
+                if (X is None) and (W is None):  # In this case predict above returns a single row
+                    Y_pred = np.tile(Y_pred, Y.shape[0])
+                    T_pred = np.tile(T_pred, T.shape[0])
+                Y_res = Y - Y_pred.reshape(Y.shape)
+                T_res = T - T_pred.reshape(T.shape)
+                return Y_res, T_res
+
+        class ModelFinal:
+            """
+            Final model at fit time, fits a residual on residual regression with a heterogeneous coefficient
+            that depends on X, i.e.
+
+                .. math ::
+                    Y - E[Y | X, W] = \\theta(X) \\cdot (T - E[T | X, W]) + \\epsilon
+
+            and at predict time returns :math:`\\theta(X)`. The score method returns the MSE of this final
+            residual on residual regression.
+            """
+
+            def __init__(self, model_final):
+                self._model_final = clone(model_final, safe=False)
+
+            def fit(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None, sample_var=None):
+                Y_res, T_res = nuisances
+                self._model_final.fit(X, T_res, Y_res, sample_weight=sample_weight, sample_var=sample_var)
+                return self
+
+            def predict(self, X=None):
+                return self._model_final.predict(X)
+
+            def score(self, Y, T, X=None, W=None, Z=None, nuisances=None, sample_weight=None, sample_var=None):

```

Comment:
sample_var is not used

Reply:
The _OrthoLearner expects this signature from the score functions of the model_final.